### ISSUE TITLE
[QUESTION] Is it possible to use other optimizer or custom optimizer?

### ISSUE THREAD

**Issue by KwangryeolPark** [NVIDIA/Megatron-LM#508]
I'm researching for improved optimzier. I have my own PyTorch optimizer code and I want to use it in Megatron-LM project.
How can I adapt my optimizer?

**Comment by deepakn94** [comment]
You could probably modify the code here to use your own optimizer: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/optimizer/__init__.py#L74.

**Comment by KwangryeolPark** [comment]
> You could probably modify the code here to use your own optimizer: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/optimizer/__init__.py#L74.

Thank you for comment. I also saw the code and edit the original Adam and SGD code to use Adam and SGD provided by PyTorch. However, while I training a model, the performance is so difficult when use optimizers from PyTorch. Are there any configurations for the same performance?

**Comment by deepakn94** [comment]
What do you mean by performance? Convergence rate or throughput?

**Comment by KwangryeolPark** [comment]
> What do you mean by performance? Convergence rate or throughput?

I mean the loss. When I use PyTorch Adam, the loss graph is so flucuated with same configuration.

**Comment by deepakn94** [comment]
I don't follow. Using the PyTorch optimizers instead of the apex optimizers gives you unstable losses? Do you have a graph that you can share?

**Comment by KwangryeolPark** [comment]
> I don't follow. Using the PyTorch optimizers instead of the apex optimizers gives you unstable losses? Do you have a graph that you can share?

Sure. Here is my research result about T5, GPT-3 and BERT model on wikitext dataset mentioned <a href="https://github.com/NVIDIA/Megatron-LM/#collecting-wikipedia-training-data">here</a>

This is my modified optimizer/__init__.py. I just change Adam to torch.optim.Adam and other configurations are same.

```python
if args.optimizer == 'adam':
        optimizer = torch.optim.Adam(param_groups,
                         lr=args.lr,
                         weight_decay=args.weight_decay,
                         betas=(args.adam_beta1, args.adam_beta2),
                         eps=args.adam_eps)
    elif args.optimizer == 'sgd':
        optimizer = torch.optim.SGD(param_groups,
                        lr=args.lr,
                        weight_decay=args.weight_decay,
                        momentum=args.sgd_momentum)
```

And below is the loss graph. As you can see, it is extremely unstable.

![W B Chart 2023  9  27  오후 12_20_45](https://github.com/NVIDIA/Megatron-LM/assets/48284967/2b73f64b-d797-4aa9-aea7-03a5f0cda043)
![W B Chart 2023  9  27  오후 12_19_55 (1)](https://github.com/NVIDIA/Megatron-LM/assets/48284967/db7614cc-8b8d-4815-a767-4cdaf2b91bc1)
![W B Chart 2023  9  27  오후 12_19_55](https://github.com/NVIDIA/Megatron-LM/assets/48284967/40c90ec9-587d-4e78-86ff-50eb13db9147)

**Comment by deepakn94** [comment]
What is step on the x-axis? The number of batches seen so far or something else?

**Comment by KwangryeolPark** [comment]
> What is step on the x-axis? The number of batches seen so far or something else?

The x-axis (Step) is the number of iterations and the y-axis is loss.

**Comment by deepakn94** [comment]
Something looks extremely wrong here...the loss should get to near 3.0 with that many steps. Do you have a graph with the apex optimizer?

**Comment by KwangryeolPark** [comment]
> Something looks extremely wrong here...the loss should get to near 3.0 with that many steps. Do you have a graph with the apex optimizer?

I deleted the result with apex (because it was intended research). However, the loss graph was stable.

**Comment by deepakn94** [comment]
I am wondering if there's an issue with your data or something else; did the loss decrease to something around 3.0?

**Comment by KwangryeolPark** [comment]
> I am wondering if there's an issue with your data or something else; did the loss decrease to something around 3.0?

I did not run many iterations but here is my GPT2 lm loss. It is very stable and the second figure is comparison.

The red graph is apex optimizer(quantized) and purple one is original torch optim
![image](https://github.com/NVIDIA/Megatron-LM/assets/48284967/c7c54b1b-e80f-450c-b4e8-8d3285850217)

![image](https://github.com/NVIDIA/Megatron-LM/assets/48284967/7d564eff-c16c-441d-807d-f757cb8da15d)

**Comment by KwangryeolPark** [comment]
There is something wrong in loss-scale graph and grad-norm graph. The loss-scale is small 1048576(quantized) vs 16384(original torch optim). And grad-norm much bigger than quantized optimizer(apex).
